Extend the PodResources API to include resources allocated by DRA

moshe010 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
It extend the kubelet PodResources API List() method to return Dynamic Resources and it add Get() to query specific pod

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
KEP: https://github.com/kubernetes/enhancements/issues/3695

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Extended the kubelet's PodResources API to include resources allocated in `ResourceClaims` via `DynamicResourceAllocation`. Additionally, added a new `Get()` method to query a specific pod for its resources.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]:  https://github.com/kubernetes/enhancements/pull/3738
```

moshe010 committed:
kube features: add KubeletPodResourcesDynamicResources and KubeletPodResourcesGet

Signed-off-by: Moshe Levi <moshele@nvidia.com>

moshe010 committed:
kubelet metrics: add pod resources get metrics

Signed-off-by: Moshe Levi <moshele@nvidia.com>

moshe010 commented:
/assign @ffromani @swatisehgal @bart0sh @klueska

I need to add unit tests and e2e tests. I would appreciate reviews to make sure we are aligned

ffromani commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -28,31 +28,41 @@ import (
 	"k8s.io/dynamic-resource-allocation/resourceclaim"
 	"k8s.io/klog/v2"
 	dra "k8s.io/kubernetes/pkg/kubelet/cm/dra/plugin"
+	"k8s.io/kubernetes/pkg/kubelet/cm/dra/state"
```
podresources API does not require state management in addition to what the resource managers would require anyway. In other words: how does DRA recover on kubelet restart?

*IF* DRA requires/benefits from this state management, I'd spin off to its own PR.

moshe010 replied:
make sense, I will fill a bug that with lose the DRA cache on kubelet restart and this cause UnprepareResources  not to be called and resources not to be clean. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/manager.go#L196. 
and all the checkpointing mechanism will be handle in that PR. 
And here I will keep only the pod resource api kep changes.

moshe010 replied:
The issue for DRA recover  kubelet restart https://github.com/kubernetes/kubernetes/issues/115902

moshe010 replied:
This PR is depended on this PR https://github.com/kubernetes/kubernetes/pull/115912

ffromani commented on pkg/kubelet/cm/dra/state/state_checkpoint.go:
```diff
@@ -60,6 +60,9 @@ type ClaimInfoState struct {
 	CdiDevices []string
+
+	// ClassName is a resource class of the claim
+	ClassName string
```
I'd squash this commit in the previous one which was adding checkpoint support (can also apply to the former comment)

klueska replied:
Agreed -- this should be part of the checkointing PR

ffromani commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -27,22 +27,24 @@ import (
 )
 
-// podResourcesServerV1alpha1 implements PodResourcesListerServer
+// podResourcesServerV implements PodResourcesListerServer
```
copy/paste mistake?

ffromani commented on pkg/kubelet/cm/fake_container_manager.go:
```diff
@@ -232,6 +232,12 @@ func (cm *FakeContainerManager) GetAllocatableMemory() []*podresourcesapi.Contai
 	return nil
 }
 
+func (cm *FakeContainerManager) GetDynamicResources(pod *v1.Pod, container *v1.Container) []*podresourcesapi.DynamicResource {
```
I'd just use empty body (just return nil) here

ffromani commented on pkg/kubelet/server/server.go:
```diff
@@ -210,10 +210,17 @@ func ListenAndServeKubeletReadOnlyServer(
 
 // ListenAndServePodResources initializes a gRPC server to serve the PodResources service
-func ListenAndServePodResources(socket string, podsProvider podresources.PodsProvider, devicesProvider podresources.DevicesProvider, cpusProvider podresources.CPUsProvider, memoryProvider podresources.MemoryProvider) {
+func ListenAndServePodResources(
```
I have another PR baking which wants to repack the arguments here in a struct because this is getting out of hand. I'll publish ASAP and perhaps we can spin off this tiny change, merge it quickly and both of us benefit

ffromani replied:
https://github.com/kubernetes/kubernetes/pull/115852/commits/9f4a3bf2721b09d3f616fb5cfa744e95442c46ab

moshe010 replied:
sure sound good

moshe010 replied:
I cherrypicked https://github.com/kubernetes/kubernetes/pull/115852/commits/71be7c60bd880606af3a97a551c18d2f7212af5f

ffromani reviewed (commented):
We need to distinguish two topics here.

1. adding checkpoint to DRA. I'm still learning my way into DRA so I can't say. I'd defer comments to DRA developers (which are already tagged) and trust their judgement. *IF* DRA would benefit from adding checkpoint, I'd do in a separate PR, but really, I trust their call here.
2. podresources API proper. Pretty much straightforward, the direction seems good with only minor things spotted so far

I'm aware podresources e2e tests need some work and I'm on it in the coming days/weeks, it will be needed to re-graduate the feature to GA on it, which is one of my goals for 1.27/28

bart0sh commented:
/triage accepted
/priority important-soon
/ok-to-test

ffromani committed:
kubelet: podresources: pack parameters in a struct

To enable rate limiting, needed for GA graduation,
we need to pass more parameters to the already crowded
`ListenAndServePodresources` function.

To tidy up a bit, pack the parameters in a helper struct,
with no intended changes in behavior.

Signed-off-by: Francesco Romani <fromani@redhat.com>

moshe010 commented:
/remove-sig testing

moshe010 commented:
/remove-sig api-machinery
/remove-sig apps
/remove-sig auth
/remove-sig autoscaling
/remove-sig cli
/remove-sig release
/remove-sig scheduling

moshe010 commented:
/remove-sig cloud-provider

moshe010 commented:
/remove-area test

moshe010 commented:
/remove-area apiserver
/remove-area cloudprovider
/remove-area code-generation
/remove-area dependency
/remove-area kubectl
/remove-area release-eng

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -27,22 +27,24 @@ import (
 // NewV1PodResourcesServer returns a PodResourcesListerServer which lists pods provided by the PodsProvider
 // with device information provided by the DevicesProvider
-func NewV1PodResourcesServer(podsProvider PodsProvider, devicesProvider DevicesProvider, cpusProvider CPUsProvider, memoryProvider MemoryProvider) v1.PodResourcesListerServer {
+func NewV1PodResourcesServer(podsProvider PodsProvider, devicesProvider DevicesProvider, cpusProvider CPUsProvider, memoryProvider MemoryProvider, draProvider DRAProvider) v1.PodResourcesListerServer {
```
Can this use the new `PodResourcesProviders` type introduced in a previous commit?

moshe010 replied:
This code is cherry-pick [1] from node: podresources: add rate limiting. see [2]. 
I am not sure why @ffromani did pass it also to  NewV1PodResourcesServer.


[1] https://github.com/kubernetes/kubernetes/pull/115852/commits/f81e542f6411ddea2a10e21a85aabc6552b2b362
[2] https://github.com/kubernetes/kubernetes/pull/115852/commits/f81e542f6411ddea2a10e21a85aabc6552b2b362#r1112351927

ffromani replied:
Just because the scope of my change was limited :) I think @klueska has a good point here, feel free to change this commit in this PR as you see fit (my PR should not delay yours), I'll happily rebase mine later.

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -27,22 +27,24 @@ import (
-func NewV1PodResourcesServer(podsProvider PodsProvider, devicesProvider DevicesProvider, cpusProvider CPUsProvider, memoryProvider MemoryProvider) v1.PodResourcesListerServer {
+func NewV1PodResourcesServer(podsProvider PodsProvider, devicesProvider DevicesProvider, cpusProvider CPUsProvider, memoryProvider MemoryProvider, draProvider DRAProvider) v1.PodResourcesListerServer {
 	return &v1PodResourcesServer{
 		podsProvider:    podsProvider,
 		devicesProvider: devicesProvider,
 		cpusProvider:    cpusProvider,
 		memoryProvider:  memoryProvider,
+		draProvider:     draProvider,
```
Can this just embed the  new `PodResourcesProviders`, i.e.:
```
return &v1PodResourcesServer{
    providers PodResourcesProviders
}
```

klueska replied:
Is there a reason this direction wasn't taken in the latest iteration?

moshe010 replied:
yes, In the pod resource api we have v1[1]  and v1alpha1 [2] structs which hold the providers relevant for the version. 
So I don't think we should keep v1 and v1alpha1  as the same struct in this case. Maybe we can do it one v1alpha1 will be removed from the code. 

[1] -https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/podresources/server_v1.go#L31-L36
[2] - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/podresources/server_v1alpha1.go#L29-L32

klueska replied:
OK. thanks for clarifying.

klueska commented on pkg/kubelet/apis/podresources/types.go:
```diff
@@ -51,3 +52,8 @@ type MemoryProvider interface {
 	GetAllocatableMemory() []*podresourcesapi.ContainerMemory
 }
+
+type DRAProvider interface {
```
We've been avoiding using `DRA` as an acronym in public APIs (even in internal code). Can you change this to `DynamicResourcesProvider`? It seems to be more consistent with the rest of the naming here anyway.

klueska commented on pkg/kubelet/apis/podresources/types.go:
```diff
@@ -51,3 +52,8 @@ type MemoryProvider interface {
 }
+
+type DRAProvider interface {
+	// GetDynamicResources returns information about dynamic resources assigend to pods and containers
```
s/assigend/assigned

klueska commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -127,10 +127,11 @@ type ContainerManager interface {
 	PodMightNeedToUnprepareResources(UID types.UID) bool
 
-	// Implements the podresources Provider API for CPUs, Memory and Devices
+	// Implements the podresources Provider API for CPUs, Memory, Devices and Dynamic Resources
```
Let's just simplify this to the following to keep it from growing unbounded.
```
Implements the PodResources Provider API
```

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -1014,6 +1014,33 @@ func (cm *containerManagerImpl) GetAllocatableMemory() []*podresourcesapi.Contai
 }
 
+func (cm *containerManagerImpl) GetDynamicResources(pod *v1.Pod, container *v1.Container) []*podresourcesapi.DynamicResource {
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DynamicResourceAllocation) &&
+		utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletPodResourcesDynamicResources) {
```
It would probably be cleaner to do:
```
if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DynamicResourceAllocation) {
	return []*podresourcesapi.DynamicResource{}
}
if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletPodResourcesDynamicResources) {
	return []*podresourcesapi.DynamicResource{}
}
...
```

klueska commented on pkg/kubelet/cm/dra/types.go:
```diff
@@ -38,4 +38,18 @@ type Manager interface {
+	GetPodDynamicResources(pod *v1.Pod, container *v1.Container) ([]*DynamicResource, error)
+}
+
+// DynamicResource contains the fileds to populate to pod resources api
+type DynamicResource struct {
+	ClassName string
+
+	ClaimName string
+
+	ClaimNamespace string
+
+	CDIDevices []string
 }
```
Is this what we've done in other components as well? Or have we just reused the equivalent of `podresources.DynamicResources` to accumulate / return this information.

moshe010 replied:
no of the provider import the podresources package. (so I wanted to be consistent) 
for device plugin there is translation method between  resource-device to podresourcesapi.ContainerDevices [1]. 
In DRA we have the claim info which is not good representation of the pod dynamic resources this is why I created DynamicResource struct. 

[1] - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/container_manager.go#L211

klueska replied:
I'd need to look at how we do it for other components, but if what you say is true, I would think we should export a type we already  know about, and leave it up to code in the podresources API to figure out how to translate it into something they can use.

moshe010 replied:
I remove the DynamicResource  struct and return claimInfo list for the Pod.

klueska commented on pkg/kubelet/kubelet.go:
```diff
@@ -2710,6 +2710,7 @@ func (kl *Kubelet) ListenAndServePodResources() {
 		Devices: kl.containerManager,
 		Cpus:    kl.containerManager,
 		Memory:  kl.containerManager,
+		DRA:     kl.containerManager,
```
This should be called `DynamicResources`

klueska commented on pkg/kubelet/server/server.go:
```diff
@@ -214,14 +214,15 @@ type PodResourcesProviders struct {
 	Devices podresources.DevicesProvider
 	Cpus    podresources.CPUsProvider
 	Memory  podresources.MemoryProvider
+	DRA     podresources.DRAProvider
```
likewise here

klueska commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -77,3 +79,31 @@ message TopologyInfo {
+
+// GetPodResourcesRequest contains information about the pod
+message GetPodResourcesRequest {
+    string name = 1;
+    string namespace = 2;
```
can we call these `pod_name` and `pod_namespace` for consistency.

klueska commented on pkg/kubelet/server/server.go:
```diff
@@ -214,14 +214,15 @@ type PodResourcesProviders struct {
 
 	podresourcesapiv1alpha1.RegisterPodResourcesListerServer(server, podresources.NewV1alpha1PodResourcesServer(providers.Pods, providers.Devices))
-	podresourcesapi.RegisterPodResourcesListerServer(server, podresources.NewV1PodResourcesServer(providers.Pods, providers.Devices, providers.Cpus, providers.Memory))
+	podresourcesapi.RegisterPodResourcesListerServer(server, podresources.NewV1PodResourcesServer(providers.Pods, providers.Devices, providers.Cpus, providers.Memory, providers.DRA))
```
This will go away if we start passing the struct instead of each individual provider.

moshe010 committed:
kubelet podresources: add unit tests for DyanmicResource and Get method

Signed-off-by: Moshe Levi <moshele@nvidia.com>

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -27,22 +27,24 @@ import (
 )
 
-// podResourcesServerV1alpha1 implements PodResourcesListerServer
+// podResourcesServerV1 implements PodResourcesListerServer
 type v1PodResourcesServer struct {
```
this comment doesn't match the type.

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -27,22 +27,24 @@ import (
-	devicesProvider DevicesProvider
-	cpusProvider    CPUsProvider
-	memoryProvider  MemoryProvider
+	podsProvider             PodsProvider
+	devicesProvider          DevicesProvider
+	cpusProvider             CPUsProvider
+	memoryProvider           MemoryProvider
+	dynamicResourcesProvider DynamicResourcesProvider
```
It's still unclear to me why we can't just include  a single `PodResourcesProviders` field here as previously suggested. Please comment.

klueska replied:
Ack

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -94,3 +96,39 @@ func (p *v1PodResourcesServer) GetAllocatableResources(ctx context.Context, req
+	pod, exist := p.podsProvider.GetPodByName(req.PodNamespace, req.PodName)
+	if !exist {
+		metrics.PodResourcesEndpointErrorsGetCount.WithLabelValues("v1").Inc()
+		return nil, fmt.Errorf("Pod %s in namespace %s not found", req.PodName, req.PodNamespace)
```
lower case

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -94,3 +96,39 @@ func (p *v1PodResourcesServer) GetAllocatableResources(ctx context.Context, req
+
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletPodResourcesGet) {
+		metrics.PodResourcesEndpointErrorsGetCount.WithLabelValues("v1").Inc()
+		return nil, fmt.Errorf("Pod Resources API Get method disabled")
```
lower case

moshe010 replied:
maybe we should keep this as upper case but change it to " PodResources API Get method disabled" same change will apply here as well https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/podresources/server_v1.go#L88

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -94,3 +96,39 @@ func (p *v1PodResourcesServer) GetAllocatableResources(ctx context.Context, req
+		metrics.PodResourcesEndpointErrorsGetCount.WithLabelValues("v1").Inc()
+		return nil, fmt.Errorf("Pod %s in namespace %s not found", req.PodName, req.PodNamespace)
+	}
+	podResources := &v1.PodResources{
```
nit: newline above this

klueska commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -94,3 +96,39 @@ func (p *v1PodResourcesServer) GetAllocatableResources(ctx context.Context, req
+		}
+	}
+
+	return &v1.GetPodResourcesResponse{
+		PodResources: podResources,
+	}, nil
```
please split into:
```
response := &v1.GetPodResourcesResponse{
	PodResources: podResources,
}

return response, nil
```

moshe010 replied:
it will be not consistent with other return calls see [1]. Do you want me to change them as well?


[1]- https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/podresources/server_v1.go#L91-L94

klueska replied:
Ugh. OK, better to stay consistent. We can change all of them as a follow-up.

klueska commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -127,10 +127,11 @@ type ContainerManager interface {
 	PodMightNeedToUnprepareResources(UID types.UID) bool
 
-	// Implements the podresources Provider API for CPUs, Memory and Devices
+	// Implements the PodResources Provider API
 	podresources.CPUsProvider
 	podresources.DevicesProvider
 	podresources.MemoryProvider
+	podresources.DynamicResourcesProvider
```
Just out of curiosity, why doesn't it implement the `Pods` provider? How is that different from the others?

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -253,3 +232,26 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 	return m.cache.hasPodReference(UID)
 }
+
+// GetPodClaimInfo gets Pod's ClaimInfo List
+func (m *ManagerImpl) GetPodClaimInfoList(pod *v1.Pod, container *v1.Container) ([]*claimInfo, error) {
```
This comment doesn't match the name of the function. Also, I'd suggest just calling it `GetPodClaimInfos()`.

Also, it is not recommended to return an unexprted type like this (i.e. `claimInfo` as lower-case). We should consider exporting it if we are going to return it out of this package like this.

klueska replied:
Or maybe the better thing is just to transform this into a function that returns a `ClaimInfoStateList` (which can then be used internally as well when we generate this list in `syncToCheckpoint()`).

moshe010 replied:
I am not sure how we can reuse it. Here we want to return the claimInfoState per Pod and in the syncToCheckpoint we don't care about what claimInfo belongs to per pod. (we just want to checkpoint all of them)

I can do as you suggested to return the ClaimInfoStateList  instead of the claimInfo itself.

ffromani commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -28,21 +28,24 @@ import (
 )
 
 // v1PodResourcesServer implements PodResourcesListerServer
+>>>>>>> 5d7dec83e0c (kubelet podresources: extend List to support Dynamic Resources and implement Get API)
```
this seems a merge conflict not fully resolved, please address

ffromani commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -965,6 +965,35 @@ func (cm *containerManagerImpl) GetAllocatableMemory() []*podresourcesapi.Contai
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DynamicResourceAllocation) {
+		return []*podresourcesapi.DynamicResource{}
+	}
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.KubeletPodResourcesDynamicResources) {
+		return []*podresourcesapi.DynamicResource{}
+	}
```
I'm not sure we need this extra check here. We should check this FG only at podresources layer, it seems weird to run this check also in the container manager layer. If the podresources api FG is not enabled this function should never be called in the first place.
If you want to be super careful, perhaps add a log on line 970 just before the return? I'd say V(3)

moshe010 replied:
It not extra check. It is to avoid doing it on the podresources layer. Do you think it better to move it there [1] ?
We will need to this on List and Get and because the call is per container. 

[1] - https://github.com/kubernetes/kubernetes/pull/115847/files#diff-4240391207d84af10161aab4338a77e780479cf90c5565d5b3f8817fbdd5efb5R129

klueska replied:
Yes, please move it up and check it in both places.

ffromani commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -965,6 +965,35 @@ func (cm *containerManagerImpl) GetAllocatableMemory() []*podresourcesapi.Contai
+	}
+	for _, podClaimInfoState := range podClaimInfoStateList {
+		for _, cdiDevice := range podClaimInfoState.CdiDevices {
+			cdiDevices = append(cdiDevices, &podresourcesapi.CDIDevice{Name: cdiDevice})
```
what if we have more than a claim? it seems to me `cdiDevices` is never reset to empty, and it probably should. Could you please check?

moshe010 replied:
I think you are right. I will check and update

klueska replied:
This is going to change with this:
https://github.com/kubernetes/kubernetes/pull/116513

Maybe we should merge that quickly so that this doesn't need to change again.

ffromani replied:
thanks both. This part LGTM now from podresources API perspective.

ffromani commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -254,3 +255,26 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
+
+// GetPodClaimInfoState gets Pod's ClaimInfoState
+func (m *ManagerImpl) GetPodClaimInfoState(pod *v1.Pod, container *v1.Container) ([]*state.ClaimInfoState, error) {
+	claimInfoState := []*state.ClaimInfoState{}
```
nit: initialize with `len(pod.Spec.ResourceClaims)` (just because you do something similar elsewhere in this PR)

ffromani commented on pkg/kubelet/cm/dra/state/state_checkpoint_test.go:
```diff
@@ -124,9 +125,10 @@ func TestCheckpointStateStore(t *testing.T) {
 	}
 
-	expectedCheckpoint := `{"version":"v1","entries":[{"DriverName":"test-driver.cdi.k8s.io","ClaimUID":"067798be-454e-4be4-9047-1aa06aea63f7","ClaimName":"example","Namespace":"default","PodUIDs":{"139cdb46-f989-4f17-9561-ca10cfb509a6":{}},"CdiDevices":["example.com/example=cdi-example"]}],"checksum":2939981547}`
+	expectedCheckpoint := `{"version":"v1","entries":[{"DriverName":"test-driver.cdi.k8s.io","ClaimUID":"067798be-454e-4be4-9047-1aa06aea63f7","ClaimName":"example","Namespace":"default","PodUIDs":{"139cdb46-f989-4f17-9561-ca10cfb509a6":{}},"CdiDevices":["example.com/example=cdi-example"],"ClassName":"class-name"}],"checksum":2939981547}`
```
can we add (here or in its proper place) a test with 2+ claims? It seems to me this should be allowed by the DRA API, either already or in the near future.

moshe010 replied:
done

klueska replied:
This comment is more relevant on:
https://github.com/kubernetes/kubernetes/pull/116513

I will add such a test

klueska replied:
Tests added in other PR

ffromani reviewed (commented):
Initial review. I will have another pass covering the tests ASAP.

moshe010 commented:
Please wait with the review I did I bad rebase. Sorry

ffromani commented on pkg/kubelet/apis/podresources/types.go:
```diff
@@ -35,6 +35,7 @@ type DevicesProvider interface {
 // PodsProvider knows how to provide the pods admitted by the node
 type PodsProvider interface {
 	GetPods() []*v1.Pod
+	GetPodByName(namespace, name string) (*v1.Pod, bool)
```
ok, this is already implemented and we just need to expose it in the `PodsProvider` interface.

moshe010 replied:
yes that true

ffromani commented on pkg/kubelet/cm/dra/state/state_checkpoint_test.go:
```diff
@@ -59,12 +59,37 @@ func TestCheckpointGetOrCreate(t *testing.T) {
 			},
 		},
+		{
+			"Restore valid checkpoint with 2 claims",
```
Looking again, this LGTM (but I haven't explored deeply yet) but does it belong in this PR? Meaning, shouldn't it be part of the PR which adds checkpoint/state support to DRA?

klueska replied:
I'm confused by the inclusion of this in this PR as well. I have already added similar tests in https://github.com/kubernetes/kubernetes/pull/116513, which this PR should be rebased on top of.

ffromani reviewed (commented):
thanks, getting there. Codewise looks good and tests looks good for now.
I'll try to run them myself to doublecheck, especially when DRA is not enabled (looking for regressions, preferably lack thereof)

moshe010 committed:
Add ClassName to chekpoint state and in-memory cache

Signed-off-by: Moshe Levi <moshele@nvidia.com>

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -57,14 +57,15 @@ type claimInfoCache struct {
 }
 
-func newClaimInfo(driverName string, claimUID types.UID, claimName, namespace string, podUIDs sets.Set[string], cdiDevice []string) (*claimInfo, error) {
+func newClaimInfo(driverName string, claimUID types.UID, claimName, namespace string, podUIDs sets.Set[string], cdiDevice []string, className string) (*claimInfo, error) {
```
yes, everyhwere

moshe010 replied:
done

moshe010 committed:
kubelet podresources: extend List to support Dynamic Resources and implement Get API

Signed-off-by: Moshe Levi <moshele@nvidia.com>

moshe010 commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -63,11 +65,21 @@ func (p *v1PodResourcesServer) List(ctx context.Context, req *v1.ListPodResource
+					Devices: p.devicesProvider.GetDevices(string(pod.UID), container.Name),
+					CpuIds:  p.cpusProvider.GetCPUs(string(pod.UID), container.Name),
+					Memory:  p.memoryProvider.GetMemory(string(pod.UID), container.Name),
+				}
```
done

moshe010 commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -94,3 +106,51 @@ func (p *v1PodResourcesServer) GetAllocatableResources(ctx context.Context, req
+				Devices: p.devicesProvider.GetDevices(string(pod.UID), container.Name),
+				CpuIds:  p.cpusProvider.GetCPUs(string(pod.UID), container.Name),
+				Memory:  p.memoryProvider.GetMemory(string(pod.UID), container.Name),
+			}
```
done

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -965,6 +965,36 @@ func (cm *containerManagerImpl) GetAllocatableMemory() []*podresourcesapi.Contai
+		return []*podresourcesapi.DynamicResource{}
+	}
+	for _, containerClaimInfo := range containerClaimInfoList {
+		// TODO currently flat list of CDIDevices we may need to change it to expose kubeletPlugin
```
nit
```suggestion
		// TODO: currently flat list of CDIDevices we may need to change it to expose kubeletPlugin
```

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -289,3 +289,26 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 	return m.cache.hasPodReference(UID)
 }
+
+// GetContainerClaimInfo gets Container's ClaimInfo
+func (m *ManagerImpl) GetContainerClaimInfo(pod *v1.Pod, container *v1.Container) ([]*ClaimInfo, error) {
```
This should be plural `GetCongtainerClaimInfos()`

klueska commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -77,3 +79,31 @@ message TopologyInfo {
+    string class_name = 1;
+    string claim_name = 2;
+    string claim_namespace = 3;
+    repeated CDIDevice cdi_devices = 4;
```
Let's wrap `CDIDevice` in another message in order to future proof it in case we decide to add the `KubeletPluginName` later

moshe010 commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -77,3 +79,35 @@ message TopologyInfo {
+    string class_name = 1;
+    string claim_name = 2;
+    string claim_namespace = 3;
+    repeated PluginCDIDevices plugins_cdi_devices = 4;
```
another option is to change this to map<string, PluginCDIDevices> plugins_cdi_devices and the string can be the kubelet plugin name.

klueska replied:
The issue is that I know of upcoming uses where resource claims will be used with different types of plugins that do not return CDIDevices at all (bother rather a different type of resource). I want to future proof for them as well in a wrapping struct that (in the future) could have a type field and may multiple optional fields depending on the type (with CDIDevices being one of them).

klueska replied:
What about:
```
message DynamicResource {
    string class_name = 1;
    string claim_name = 2;
    string claim_namespace = 3;
    repeated ClaimResource claim_resources = 4;
}

message ClaimResource {
    repeated CDIDevice cdi_devices = 1;
    // add other possible fields here in the future
}
```

moshe010 replied:
So something like this:
message DynamicResource {
    string class_name = 1;
    string claim_name = 2;
    string claim_namespace = 3;
    repeated Resources resources = 4;
}

message Resources {
  oneof resource{
        CDIDevices cdi_devices = 1;
  }    
}

message CDIDevices {
  repeated CDIDevice cdi_devices = 1;
}

klueska replied:
`oneof` is not allowed in core k8s code

klueska replied:
You can find it in vendored code, but not within core kubernetes itself.

klueska replied:
Actually, I can't find any reference that says it's explicitly disallowed, I just don't know of any code that uses it (and a simple grep turns up nothing).

moshe010 replied:
by the way the ClaimResource  is per plugin as I doing https://github.com/kubernetes/kubernetes/pull/115847/files/d5fc1e401a80ccc69f2a7318d08621f57c48cddf#diff-c07d343a8d459721656cf7ddacda71506bde01a5cac288d15a4aba6bfc9a1caeR987 ?

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -965,6 +965,39 @@ func (cm *containerManagerImpl) GetAllocatableMemory() []*podresourcesapi.Contai
+	}
+	for _, containerClaimInfo := range containerClaimInfos {
+		var claimResources []*podresourcesapi.ClaimResource
+		// TODO: Currently a flat list of ClaimResource; we may need to change it to expose the kubelet plugin name
```
```suggestion
// TODO: Currently we maintain a list of ClaimResources, each of which contains a set of CDIDevices from a different kubelet plugin. In the future we may want to include the name of the kubelet plugin and/or other types of resources that are not CDIDevices (assuming the DRAmanager supports this).
```

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -965,6 +965,39 @@ func (cm *containerManagerImpl) GetAllocatableMemory() []*podresourcesapi.Contai
+			for _, cdiDevice := range klPluginCdiDevices {
+				cdiDevices = append(cdiDevices, &podresourcesapi.CDIDevice{Name: cdiDevice})
+			}
+			claimResources = append(claimResources, &podresourcesapi.ClaimResource{CdiDevices: cdiDevices})
```
```suggestion
			claimResources = append(claimResources, &podresourcesapi.ClaimResource{CDIDevices: cdiDevices})
```

klueska commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -77,3 +79,36 @@ message TopologyInfo {
+
+// ClaimResource contains per plugin resource information
+message ClaimResource {
+    repeated CDIDevice cdi_devices = 1;
```
```suggestion
    repeated CDIDevice cdi_devices = 1 [(gogoproto.customname) = "CDIDevices"];
```

moshe010 replied:
nice I wasn't aware of this

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -288,3 +289,26 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
+			if podResourceClaim.Name != claim.Name {
+				continue
+			}
+
```
nit. remove newline

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -288,3 +289,26 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
+			claimInfos = append(claimInfos, claimInfo)
+		}
+	}
+
```
nit: remove newline

klueska reviewed (commented):
Mostly cosmetic comments at this point. Haven't looked at the tests yet though. @ffromani is doing a first pass.

ffromani commented:
do we need the windows lane here?

ffromani commented:
/test pull-kubernetes-unit

ffromani commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -80,76 +110,76 @@ func TestListPodResourcesV1(t *testing.T) {
+								Devices:          []*podresourcesapi.ContainerDevices{},
+								CpuIds:           cpus,
+								Memory:           memory,
+								DynamicResources: draDevs,
```
I'd add a testcase with both dynamic resources and devices.

ffromani replied:
can be postponed.

ffromani commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -171,12 +202,14 @@ func TestListPodResourcesV1(t *testing.T) {
 			mockCPUsProvider.EXPECT().GetAllocatableCPUs().Return([]int64{}).AnyTimes()
 			mockDevicesProvider.EXPECT().GetAllocatableDevices().Return([]*podresourcesapi.ContainerDevices{}).AnyTimes()
 			mockMemoryProvider.EXPECT().GetAllocatableMemory().Return([]*podresourcesapi.ContainerMemory{}).AnyTimes()
+			mockDynamicResourcesProvider.EXPECT().GetDynamicResources(pods[0], &containers[0]).Return([]*podresourcesapi.DynamicResource{}).AnyTimes()
```
shouldn't this return `tc.dynamicResources` ? the reason why the test is still passing seems to be you didn't yet update `equalListResponse` below to check the dynamic resources

ffromani commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -485,6 +518,185 @@ func TestAllocatableResources(t *testing.T) {
+					},
+				},
+			},
+		},
```
similarly to above, I think we want a test with both devices and dynamic resources

ffromani replied:
can be postponed.

ffromani commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -485,6 +518,185 @@ func TestAllocatableResources(t *testing.T) {
+			mockCPUsProvider.EXPECT().GetAllocatableCPUs().Return([]int64{}).AnyTimes()
+			mockDevicesProvider.EXPECT().GetAllocatableDevices().Return([]*podresourcesapi.ContainerDevices{}).AnyTimes()
+			mockMemoryProvider.EXPECT().GetAllocatableMemory().Return([]*podresourcesapi.ContainerMemory{}).AnyTimes()
+			mockDynamicResourcesProvider.EXPECT().GetDynamicResources(pod, &containers[0]).Return([]*podresourcesapi.DynamicResource{}).AnyTimes()
```
twice? the second instance looks wrong, the previous correct

ffromani commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -581,3 +793,33 @@ func equalAllocatableResourcesResponse(respA, respB *podresourcesapi.Allocatable
+			return false
+		}
+
+		if !equalContainerDevices(cntA.Devices, cntB.Devices) {
```
you seem not to check for dynamic resources here, this could explain why the suspicious tests above (double mock registration, the latter possibly wrong) still passes

ffromani reviewed (commented):
the unit tests needs a bit of extra work

ffromani commented:
please rebase

LGTM, all comments addressed

ffromani commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -515,11 +753,54 @@ func equalListResponse(respA, respB *podresourcesapi.ListPodResourcesResponse) b
 				return false
 			}
+
+			if !euqalDynamicResources(cntA.DynamicResources, cntB.DynamicResources) {
```
let's fix the typo (cosmetic only, because the typo is used consistently!) before beta. Not blocking now.

ffromani reviewed (commented):
/lgtm

final LGTM call is @klueska 's

needs approval for the feature gate part

ffromani commented:
/hold

because:

> /lgtm
> 
> final LGTM call is @klueska 's

ffromani commented:
/retest

ffromani commented:
/test pull-kubernetes-unit

unrelated failure

klueska commented:
Thanks for all your hard work on this @moshe010
/lgtm
/approve

klueska commented:
/release-note-edit
```
release-note
We extend the kubelet's PodResources API to include resources allocated in `ResourceClaims` via `DynamicResourceAllocation`. Additionally, we add a new `Get()` method to query a specific pod for its resources rather than forcing users to rely solely on the existing `List()` method which returns resources for all pods.
```

klueska commented:
/release-note-edit

```release-note
We extend the kubelet's PodResources API to include resources allocated in `ResourceClaims` via `DynamicResourceAllocation`. Additionally, we add a new `Get()` method to query a specific pod for its resources rather than forci

klueska commented:
/retest

klueska commented:
/cc dchen1107 derekwaynecarr
For feature gate approval

klueska commented:
/assign dchen1107 derekwaynecarr
For feature gate approval

klueska commented:
/lgtm
/approve

ffromani commented:
/hold cancel

Kevin reviewed

dims commented:
/approve

new alpha feature gates look good to me!

dchen1107 commented:
/approve the feature gate

moshe010 commented:
follow-up PR https://github.com/kubernetes/kubernetes/pull/116742 to address the no-blocking issues
